include/mbgl/style/expression/literal.hpp

```diff
@@ -12,8 +12,18 @@ namespace expression {
 
 class Literal : public Expression {
 public:
-    Literal(Value value_) : Expression(typeOf(value_)), value(value_) {}
-    Literal(type::Array type_, std::vector<Value> value_) : Expression(type_), value(value_) {}
+    Literal(Value value_, optional<mbgl::Value> serialized_ = optional<mbgl::Value>())
+        : Expression(typeOf(value_))
+        , value(value_)
+        , serialized(std::move(serialized_))
+    {}
+    
+    Literal(type::Array type_, std::vector<Value> value_, optional<mbgl::Value> serialized_ = optional<mbgl::Value>())
+        : Expression(type_)
+        , value(value_)
+        , serialized(std::move(serialized_))
+    {}
```

Comment:
I'd say let's not add this extra bit of storage / complexity for now, and just serialized folded literals as-is.